test/Dialect/Arc/lower-state.mlir

```diff
@@ -1,296 +1,611 @@
-  %0 = arc.memory_read_port %mem[%c0_i2] : <4 x i42, i2>
-  // CHECK-NEXT: return
-  arc.output %0 : i42
+  // CHECK:   [[TMP:%.+]] = arc.call @IdI42Arc([[READ_OLD]])
+  // CHECK:   arc.state_write [[Q0]] = [[TMP]]
+  %q0 = arc.state @IdI42Arc(%0) clock %clock latency 1 {names = ["q0"]} : (i42) -> i42
+  // CHECK: }
+  // CHECK: [[Q0_NEW:%.+]] = arc.state_read [[Q0]]
+  // CHECK: arc.state_write %out_c = [[Q0_NEW]]
+  hw.output %q0 : i42
 }
 
-// CHECK-LABEL:  arc.model @maskedMemoryWrite
-hw.module @maskedMemoryWrite(in %clk: !seq.clock) {
-  %true = hw.constant true
-  %c0_i2 = hw.constant 0 : i2
-  %c9001_i42 = hw.constant 9001 : i42
-  %c1010_i42 = hw.constant 1010 : i42
+// CHECK-LABEL: arc.model @MemoryEnable
+hw.module @MemoryEnable(in %clock: !seq.clock, in %a: i2, in %b: i42, in %en: i1) {
+  // CHECK: [[MEM:%.+]] = arc.alloc_memory
+  // CHECK: scf.if {{%.+}} {
+  // CHECK:   [[A:%.+]] = arc.state_read %in_a
+  // CHECK:   [[B:%.+]] = arc.state_read %in_b
+  // CHECK:   [[EN:%.+]] = arc.state_read %in_en
+  // CHECK:   [[TMP:%.+]]:3 = arc.call @IdI2AndI42AndI1Arc([[A]], [[B]], [[EN]])
+  // CHECK:   scf.if [[TMP]]#2 {
+  // CHECK:     arc.memory_write [[MEM]][[[TMP]]#0], [[TMP]]#1
+  // CHECK:   }
+  // CHECK: }
   %mem = arc.memory <4 x i42, i2>
-  arc.memory_write_port %mem, @identity2(%c0_i2, %c9001_i42, %true, %c1010_i42) clock %clk enable mask latency 1 : <4 x i42, i2>, i2, i42, i1, i42
+  arc.memory_write_port %mem, @IdI2AndI42AndI1Arc(%a, %b, %en) clock %clock enable latency 1 : <4 x i42, i2>, i2, i42, i1
 }
-arc.define @identity2(%arg0: i2, %arg1: i42, %arg2: i1, %arg3: i42) -> (i2, i42, i1, i42) {
-  arc.output %arg0, %arg1, %arg2, %arg3 : i2, i42, i1, i42
+
+// CHECK-LABEL: arc.model @MemoryEnableAndMask
+hw.module @MemoryEnableAndMask(in %clock: !seq.clock, in %a: i2, in %b: i42, in %en: i1, in %mask: i42) {
+  // CHECK: [[MEM:%.+]] = arc.alloc_memory
+  // CHECK: scf.if {{%.+}} {
+  // CHECK:   [[A:%.+]] = arc.state_read %in_a
+  // CHECK:   [[B:%.+]] = arc.state_read %in_b
+  // CHECK:   [[EN:%.+]] = arc.state_read %in_en
+  // CHECK:   [[MASK:%.+]] = arc.state_read %in_mask
+  // CHECK:   [[TMP:%.+]]:4 = arc.call @IdI2AndI42AndI1AndI42Arc([[A]], [[B]], [[EN]], [[MASK]])
+  // CHECK:   scf.if [[TMP]]#2 {
+  // CHECK:     [[ALL_ONES:%.+]] = hw.constant -1
+  // CHECK:     [[MASK_INV:%.+]] = comb.xor bin [[TMP]]#3, [[ALL_ONES]]
+  // CHECK:     [[DATA_OLD:%.+]] = arc.memory_read [[MEM]][[[TMP]]#0]
+  // CHECK:     [[MASKED_OLD:%.+]] = comb.and bin [[MASK_INV]], [[DATA_OLD]]
+  // CHECK:     [[MASKED_NEW:%.+]] = comb.and bin [[TMP]]#3, [[TMP]]#1
+  // CHECK:     [[DATA_NEW:%.+]] = comb.or bin [[MASKED_OLD]], [[MASKED_NEW]]
+  // CHECK:     arc.memory_write [[MEM]][[[TMP]]#0], [[DATA_NEW]]
+  // CHECK:   }
+  // CHECK: }
+  %mem = arc.memory <4 x i42, i2>
+  arc.memory_write_port %mem, @IdI2AndI42AndI1AndI42Arc(%a, %b, %en, %mask) clock %clock enable mask latency 1 : <4 x i42, i2>, i2, i42, i1, i42
```

Comment:
A second masked write to the same location, but inverted mask could be interesting to check the correct 'old' value is read.

Reply:
Yeah we should definitely have that as an integration test somewhere. I don't think the correct behavior would be apparent here in the lowering test :thinking: